modules/roles/server/controllers/roles.controller.js

```diff
@@ -31,191 +47,255 @@ function roleModule(logger, userModule) {
 
 
   /*
-  * Roles can be added to the role tree at any level other than root.
-  * When a role is added, the tree is traversed, and for each 'parent'
-  * the role has, any users with role equal to that parent role will 
-  * get the new role in their subroles array.
-  * 
-  * @param {req}  The Express HTTP request with the role in the body
-  * @param {res} The Express HTTP response object
-  * @param {next} The Express HTTP next function
-  *
-  * @return {bool} whether or not the role was added
-  */
-  function addRole(req, res, next) 
+   * Roles can be added to the role tree at any level other than root.
+   * When a role is added, the tree is traversed, and for each 'parent'
+   * the role has, any users with role equal to that parent role will 
+   * get the new role in their subroles array.
+   * 
+   * @param {req}  The Express HTTP request with the role in the body
+   * @param {res} The Express HTTP response object
+   * @param {next} The Express HTTP next function
+   *
+   * @return {bool} whether or not the role was added
+   */
+  function addRole(req, res, next)
   {
-    var role = req.body;
-    //make sure the role is valid
-    if(role._id != null)
+    //validation check
+    if (!req.body._id || !req.body.parent)
     {
-      var addedRole = new Roles();
-      addedRole._id = role._id;
-      addedRole.parent = role.parent || null;
-      role.canModify = role.canModify || false;
-      role.parentForDescendants = role.parentForDescendants || [];
-
-      //check to make sure the parent exists
-      if(addedRole.parent === null || !roleExists(addedRole.parent))
-      {
-        res.status(500).send("Parent must be a valid role");
-        return;
-      }
-
-
-      //save will fail if _id isn't unique
-      addedRole.save((err, data) => {
-        if(err)
+      return sendServerError(res, "required role fields not set");
```

Comment:
I like the abstraction, but you don't pass a code with this method.  That's a pretty import piece of a error response.